### PR TITLE
Campaign Placement Metadata Hierarchy

### PR DESCRIPTION
**Pull request by raileanu-adobe**
Please link to the issue #…

### CODE REVIEW

File: extensions/adobe/experience/adcloud/dsp/campaign.schema.json [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+          "format": "date-time",
+          "description": "The insertion order fields representing campaign start date."
+        },
+        "xdm:campaignEndTime": {
+          "title": "Campaign end date.",
+          "type": "string",
+          "format": "date-time",
+          "description": "The insertion order fields representing campaign end date."
+        },
+        "xdm:accountId": {
```
Comment (udayeyunni) [inline extensions/adobe/experience/adcloud/dsp/campaign.schema.json]: Redundant in Campaign schemas but I am OK if you would like to include.

File: extensions/adobe/experience/adcloud/dsp/campaign.schema.json [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+          "title": "Account Identifier",
+          "type": "string",
+          "description": "Identifier for the account owning this campaign."
+        },
+        "xdm:advertiserId": {
+          "title": "Advertiser Identifier",
+          "type": "string",
+          "description": "Identifier for the advertiser owning this campaign."
+        },
+        "xdm:packageId": {
```
Comment (udayeyunni) [inline extensions/adobe/experience/adcloud/dsp/campaign.schema.json]: Campaigns are not associated with Packages in DSP. I don't see a field package_id in insertion_order table.

File: extensions/adobe/experience/adcloud/dsp/campaign.schema.json [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+          "enum": [
+            "Media",
+            "Brandsights"
+          ],
+          "meta:enum": {
+            "Media": "Media",
+            "Brandsights": "Brandsights"
+          }
+        },
+        "xdm:campaignStartTime": {
```
Comment (udayeyunni) [inline extensions/adobe/experience/adcloud/dsp/campaign.schema.json]: Confirm with DSP team insertion_order table is the correct table where Start and End Times are stored. There doesn't seem to be 1 to 1 mapping between insertion_order and campaign tables. Not sure when insertion_order made into Promo2. Stats team can answer this question.

File: extensions/adobe/experience/adcloud/dsp/campaign.schema.json [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    },
+    {
+      "$ref": "https://ns.adobe.com/xdm/common/auditable"
+    },
+    {
+      "$ref": "#/definitions/dsp-campaign"
+    }
+  ],
+  "meta:status": "experimental"
+}
```
Comment (udayeyunni) [inline extensions/adobe/experience/adcloud/dsp/campaign.schema.json:122]: Recommend including the fields 'user_id' (not sure whether we have to filter by this field; Someone in STATS team should know)

File: extensions/adobe/experience/adcloud/dsp/package.schema.json [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        "xdm:unitsPurchased": {
+          "title": "Units Purchased",
+          "type": "number",
+          "description": "The number of purchased units."
+        },
+        "xdm:goalMetric": {
+          "title": "Goal Metric",
+          "type": "string",
+          "description": "The metric on which the pacing occurs so that the campaign hits the desired metric value over the course of its lifetime. E.g. SPEND, IMPRESSIONS"
+        },
```
Comment (udayeyunni) [inline extensions/adobe/experience/adcloud/dsp/package.schema.json:54]: Check whether it makes sense to define goalMetric as enum or data type.

File: extensions/adobe/experience/adcloud/dsp/package.schema.json [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        "xdm:goalMetricValue": {
+          "title": "Goal Interval",
+          "type": "integer",
+          "description": "The maximum value allowed for the metric to reach in a certain interval."
+        },
+        "xdm:goalType": {
+          "title": "Goal Type",
+          "type": "string",
+          "description": "The action taken when the metric hits the maximum value. E.g. CAP, PACE "
+        },
```
Comment (udayeyunni) [inline extensions/adobe/experience/adcloud/dsp/package.schema.json:72]: See comment for goalMetric

File: extensions/adobe/experience/adcloud/dsp/placement.schema.json [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    "This work is licensed under a Creative Commons Attribution 4.0 International (CC BY 4.0) license",
+    "you may not use this file except in compliance with the License. You may obtain a copy",
+    "of the License at https://creativecommons.org/licenses/by/4.0/"
+  ],
+  "$id": "https://ns.adobe.com/experience/adcloud/dsp/placement",
+  "$schema": "http://json-schema.org/draft-06/schema#",
+  "title": "Adobe Advertising Cloud DSP Campaign Placement",
+  "type": "object",
+  "meta:extends": ["https://ns.adobe.com/xdm/data/record"],
+  "description": "Adobe Advertising Cloud DSP Campaign Placement hierarchy details such as name, type, campaignId, account, package, advertiser, account.",
```
Comment (udayeyunni) [inline extensions/adobe/experience/adcloud/dsp/placement.schema.json]: Update the description to reflect it is DSP placement class

**Review by udayeyunni — COMMENTED** [review]
Ana, Minor edits to the schema are required. Please review and reach out if you have any further questions.

File: extensions/adobe/experience/adcloud/dsp/account.schema.json [hunk 8]
```diff
@@ -0,0 +1,92 @@
+{
+  "meta:license": [
+    "Copyright 2019 Adobe Systems Incorporated. All rights reserved.",
+    "This work is licensed under a Creative Commons Attribution 4.0 International (CC BY 4.0) license",
+    "you may not use this file except in compliance with the License. You may obtain a copy",
+    "of the License at https://creativecommons.org/licenses/by/4.0/"
+  ],
+  "$id": "https://ns.adobe.com/experience/adcloud/dsp/account",
+  "$schema": "http://json-schema.org/draft-06/schema#",
+  "title": "Adobe Advertising Cloud DSP Account",
```
Comment (udayeyunni) [inline extensions/adobe/experience/adcloud/dsp/account.schema.json]: XDM Team recommends not to include company names in title. You can change the title to "DSP Advertising ..." and see what comments we get.

File: extensions/adobe/experience/adcloud/dsp/account.example.1.json [hunk 9]
```diff
@@ -0,0 +1,10 @@
+{
+  "xdm:id": "12",
+  "xdm:accountKey": "dsdfweW12",
+  "xdm:accountName": "Dailymotion",
+  "xdm:accountType": "Brand Direct No Agency",
```
Comment (is-slo-and) [inline extensions/adobe/experience/adcloud/dsp/account.example.1.json]: The enum list value for accountType doesn't have spaces.
I'm not sure if there is any automated validation of these example files so this may not matter.

Comment (is-slo-and) [reply extensions/adobe/experience/adcloud/dsp/account.example.1.json]: Ignore my comment above...I didn't notice the meta:enum section.

File: extensions/adobe/experience/adcloud/dsp/package.schema.json (hunk 6 above)
Comment (raileanu-adobe) [reply extensions/adobe/experience/adcloud/dsp/package.schema.json:72]: An enum makes sense here because there are only a few options for it (CAP or PACE).

File: extensions/adobe/experience/adcloud/dsp/package.schema.json (hunk 5 above)
Comment (raileanu-adobe) [reply extensions/adobe/experience/adcloud/dsp/package.schema.json:54]: Not really because we have a lot of possible values in the DB (29).

File: extensions/adobe/experience/adcloud/dsp/account.schema.json (hunk 8 above)
Comment (raileanu-adobe) [reply extensions/adobe/experience/adcloud/dsp/account.schema.json]: Updated

File: extensions/adobe/experience/adcloud/dsp/placement.schema.json (hunk 7 above)
Comment (raileanu-adobe) [reply extensions/adobe/experience/adcloud/dsp/placement.schema.json]: Updated

File: extensions/adobe/experience/adcloud/dsp/campaign.schema.json (hunk 4 above)
Comment (raileanu-adobe) [reply extensions/adobe/experience/adcloud/dsp/campaign.schema.json:122]: Updated

File: extensions/adobe/experience/adcloud/dsp/campaign.schema.json (hunk 3 above)
Comment (raileanu-adobe) [reply extensions/adobe/experience/adcloud/dsp/campaign.schema.json]: Confirmed that ,when creating a campaign, the start and end dates go into insertion_order table

File: extensions/adobe/experience/adcloud/dsp/campaign.schema.json (hunk 2 above)
Comment (raileanu-adobe) [reply extensions/adobe/experience/adcloud/dsp/campaign.schema.json]: Yes, you are right. I've reversed the relationship to properly reflect the DB model : removed the "packageId" field from "campaign" and added "campaignId" field (will be taken from insertion_order table) to "package".

File: extensions/adobe/experience/adcloud/dsp/advertiser.schema.json [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        },
+        "xdm:advertiserName": {
+          "title": "Advertiser Name",
+          "type": "string",
+          "description": "The name of the advertiser."
+        },
+        "xdm:advertiserStatus": {
+          "title": "Advertiser Status",
+          "type": "string",
+          "description": "Managed Flag indicates whether advertiser is active.",
```
Comment (udayeyunni) [inline extensions/adobe/experience/adcloud/dsp/advertiser.schema.json]: Managed Flag?

File: extensions/adobe/experience/adcloud/dsp/campaign.schema.json [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+            "Brandsights"
+          ],
+          "meta:enum": {
+            "Media": "Media",
+            "Brandsights": "Brandsights"
+          }
+        },
+        "xdm:userId": {
+          "title": "User Identifier",
+          "type": "integer",
```
Comment (udayeyunni) [inline extensions/adobe/experience/adcloud/dsp/campaign.schema.json:67]: Rest of the identifiers are defined as of type "string". We should be OK with integer but I would like to flag this. Evaluate whether "string" data type makes sense for userId.

**Review by udayeyunni — APPROVED** [review]
Ana, Please take care of the minor changes and you are good to go as far as I am concerned. Thx

File: extensions/adobe/experience/adcloud/dsp/advertiser.schema.json (hunk 10 above)
Comment (raileanu-adobe) [reply extensions/adobe/experience/adcloud/dsp/advertiser.schema.json]: updated description

File: extensions/adobe/experience/adcloud/dsp/campaign.schema.json (hunk 11 above)
Comment (raileanu-adobe) [reply extensions/adobe/experience/adcloud/dsp/campaign.schema.json:67]: userId is our internal DB identifier, not mapped to any record:id so ,I assume, no need to convert it to string.

File: extensions/adobe/experience/adcloud/dsp/package.schema.json [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+  ],
+  "$id": "https://ns.adobe.com/experience/adcloud/dsp/package",
+  "$schema": "http://json-schema.org/draft-06/schema#",
+  "title": "DSP Advertising Campaign Package",
+  "type": "object",
+  "meta:extensible": true,
+  "meta:abstract": true,
+  "meta:auditable": true,
+  "meta:extends": ["https://ns.adobe.com/xdm/data/record"],
+  "description": "Adobe Advertising Cloud DSP Campaign Package Details.",
```
Comment (skannan-adobe) [inline extensions/adobe/experience/adcloud/dsp/package.schema.json]: can you remove `Campaign` in the description and title, since this is not necessarily tied to a package ?

File: extensions/adobe/experience/adcloud/dsp/package.schema.json [hunk 13]
```diff
@@ -0,0 +1,104 @@
+{
+  "meta:license": [
+    "Copyright 2019 Adobe Systems Incorporated. All rights reserved.",
+    "This work is licensed under a Creative Commons Attribution 4.0 International (CC BY 4.0) license",
+    "you may not use this file except in compliance with the License. You may obtain a copy",
+    "of the License at https://creativecommons.org/licenses/by/4.0/"
+  ],
+  "$id": "https://ns.adobe.com/experience/adcloud/dsp/package",
```
Comment (skannan-adobe) [inline extensions/adobe/experience/adcloud/dsp/package.schema.json]: this seems to have fields from `promo2.package` and `promo2.package_goal` and `promo2.goal_interval`, `promo2.goal_metric`and `promo2.goal_type` tables. Am i missing something ?

File: extensions/adobe/experience/adcloud/dsp/package.schema.json [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+            "PACE"
+          ],
+          "meta:enum": {
+            "CAP": "CAP",
+            "PACE": "PACE"
+          }
+        },
+        "xdm:goalInterval": {
+          "title": "Goal Interval",
+          "type": "string",
```
Comment (skannan-adobe) [inline extensions/adobe/experience/adcloud/dsp/package.schema.json:75]: you can add `enum` for this one too. Based on the example and the values in the `promo2.goal_interval` table, we only have a finite set of values; `ALLTIME `, `DAY`, `HOUR`, `MONTH` and `WEEK`

File: extensions/adobe/experience/adcloud/dsp/campaign.schema.json [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+          "description": "Identifier for the user who created the campaign."
+        },
+        "xdm:campaignStartTime": {
+          "title": "Campaign start date.",
```
Comment (skannan-adobe) [inline extensions/adobe/experience/adcloud/dsp/campaign.schema.json]: nitpicking here, can you use `time` instead of `date` in the title and description, just to be consistent with the name ?

File: extensions/adobe/experience/adcloud/dsp/campaign.schema.json [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+          "type": "string",
+          "format": "date-time",
+          "description": "The insertion order fields representing campaign start date."
+        },
+        "xdm:campaignEndTime": {
+          "title": "Campaign end date.",
```
Comment (skannan-adobe) [inline extensions/adobe/experience/adcloud/dsp/campaign.schema.json]: nitpicking here, can you use `time` instead of `date` in the title and description, just to be consistent with the name ?

File: extensions/adobe/experience/adcloud/dsp/placement.schema.json [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+          }
+        },
+        "xdm:placementStartTime": {
+          "title": "Start Date and Time",
+          "type": "string",
+          "format": "date-time",
+          "description": "The date on which a placement starts showing ads (ms since Jan 1, 1970)"
+          },
+        "xdm:placementEndTime": {
+          "title": "End Date and Time",
```
Comment (skannan-adobe) [inline extensions/adobe/experience/adcloud/dsp/placement.schema.json]: `Campaign Placement End Date and Time`

File: extensions/adobe/experience/adcloud/dsp/placement.schema.json [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+        },
+        "xdm:placementBudget": {
+          "title": "Campaign Placement Budget",
+          "type": "integer",
+          "description": "Campaign placement budget."
+        },
+        "xdm:campaignId": {
+          "title": "Campaign Identifier",
+          "type": "string",
+          "description": "Campaign Identifier on the display advertising platform corresponding to this placement."
```
Comment (skannan-adobe) [inline extensions/adobe/experience/adcloud/dsp/placement.schema.json]: nitpicking here; Please rephrase this to include `to which this placement belongs` or phrase it like the description used for `packageId` below;
`Identifier of the campaign this placement belongs to.`

**Review by skannan-adobe — COMMENTED** [review]
left some review comments

File: extensions/adobe/experience/adcloud/dsp/package.schema.json (hunk 12 above)
Comment (raileanu-adobe) [reply extensions/adobe/experience/adcloud/dsp/package.schema.json]: updated

File: extensions/adobe/experience/adcloud/dsp/package.schema.json (hunk 13 above)
Comment (raileanu-adobe) [reply extensions/adobe/experience/adcloud/dsp/package.schema.json]: Yes, there is also  'package_goal' table involved , but not promo2.goal_interval, promo2.goal_metric and promo2.goal_type tables. I added all these extra fields because they seemed relevant in the context of a package entity. Do you agree ? I've documented it [here](https://wiki.corp.adobe.com/display/EfficientFrontier/%5BXDM+Mapping%5D+Campaign+Placement-related+hierarchy#id-[XDMMapping]CampaignPlacement-relatedhierarchy-Package)

File: extensions/adobe/experience/adcloud/dsp/package.schema.json (hunk 14 above)
Comment (raileanu-adobe) [reply extensions/adobe/experience/adcloud/dsp/package.schema.json:75]: updated

File: extensions/adobe/experience/adcloud/dsp/campaign.schema.json (hunk 15 above)
Comment (raileanu-adobe) [reply extensions/adobe/experience/adcloud/dsp/campaign.schema.json]: updated

File: extensions/adobe/experience/adcloud/dsp/campaign.schema.json (hunk 16 above)
Comment (raileanu-adobe) [reply extensions/adobe/experience/adcloud/dsp/campaign.schema.json]: updated

File: extensions/adobe/experience/adcloud/dsp/placement.schema.json (hunk 17 above)
Comment (raileanu-adobe) [reply extensions/adobe/experience/adcloud/dsp/placement.schema.json]: updated

File: extensions/adobe/experience/adcloud/dsp/placement.schema.json (hunk 18 above)
Comment (raileanu-adobe) [reply extensions/adobe/experience/adcloud/dsp/placement.schema.json]: updated
